### ISSUE TITLE
[Zig] Implement All Evaluation Frameworks

### ISSUE THREAD

**Issue by scttfrdmn** [scttfrdmn/agenkit#310]
## Description

Implement complete evaluation framework for Zig to achieve parity with Python, Go, TypeScript, C++, and Rust.

Zig currently has **0/12 frameworks (0% complete)**. This is a large effort to build the entire evaluation infrastructure from scratch.

### ✅ Other Languages
- Python: 100% (12/12) - Reference implementation
- Go: 100% (12/12)
- TypeScript: 100% (12/12)
- C++: 100% (12/12) - Just completed
- Rust: 83% (10/12) - Missing A/B Testing, Bayesian Opt

### ❌ Zig: 0% (0/12)
No `/src/evaluation` directory exists.

#### 1. Core Module (`core.zig`)
**Purpose:** Foundation types and interfaces
**LOC:** ~300

**Key Types:**
```zig
pub const Metric = struct {
    name: []const u8,
    measure: *const fn(agent: *Agent, input: Message, output: Message, context: anytype) f64,
    aggregate: *const fn(measurements: []const f64) std.StringHashMap(f64),
};

pub const TestCase = struct {
    input: []const u8,
    expected: union(enum) {
        exact: []const u8,
        functional: *const fn([]const u8) bool,
    },
    metadata: std.StringHashMap(Value),
    tags: []const []const u8,
};

pub const EvaluationResult = struct {
    session_id: []const u8,
    n_cases: usize,
    n_passed: usize,
    metrics: std.StringHashMap(f64),
    errors: std.ArrayList(ErrorRecord),
};
```

#### 2. Metrics Collection (`metrics.zig`)
**Purpose:** Core metrics infrastructure
**LOC:** ~450

**Features:**
- Session tracking
- Metric aggregation
- MetricsCollector for cross-session stats
- JSON serialization

#### 3. Quality Metrics (`quality_metrics.zig`)
**Purpose:** Standard evaluation metrics
**LOC:** ~400

**Metrics:**
- Accuracy
- Precision/Recall
- F1 Score
- Quality scoring

#### 4. Recorder (`recorder.zig`)
**Purpose:** Session recording and replay
**LOC:** ~550

**Features:**
- Record agent sessions
- Replay for debugging
- JSON export/import

#### 5. Regression Detection (`regression.zig`)
**Purpose:** Detect performance regressions
**LOC:** ~450

**Features:**
- Compare session results
- Severity levels (CRITICAL, MAJOR, MINOR, NEGLIGIBLE)
- Threshold-based detection

#### 6. Context Metrics (`context_metrics.zig`)
**Purpose:** Extreme-scale evaluation
**LOC:** ~500

**Features:**
- Track context length (1K - 25M tokens)
- Compression ratio measurement
- Latency percentiles (p50, p90, p95, p99)

#### 7. Benchmarks (`benchmarks.zig`)
**Purpose:** Standardized test suites
**LOC:** ~600

**Benchmarks:**
- SimpleQA (5 test cases)
- NeedleInHaystack
- InformationRetention
- ExtremeScale (1M/10M/25M tokens)

#### 8. Optimizer Base (`optimizer.zig`)
**Purpose:** Hyperparameter optimization foundation
**LOC:** ~450

**Features:**
- SearchSpace (continuous, discrete, integer, categorical)
- RandomSearchOptimizer
- OptimizationResult tracking

#### 9. A/B Testing (`ab_testing.zig`)
**Purpose:** Statistical significance testing
**LOC:** ~600

**Features:**
- t-test, Mann-Whitney, Chi-square, Bootstrap
- Cohen's d (effect size)
- Confidence intervals
- Sample size calculator

#### 10. Bayesian Optimization (`bayesian_optimizer.zig`)
**Purpose:** Intelligent hyperparameter search
**LOC:** ~800

**Features:**
- Acquisition functions (EI, UCB, PI)
- Simplified surrogate model
- Configuration similarity metric
- Statistical functions (norm_cdf, norm_pdf)

#### 11. Prompt Optimization (`prompt_optimizer.zig`)
**Purpose:** Automated prompt engineering
**LOC:** ~600

**Features:**
- Grid search (exhaustive)
- Random search (efficient)
- Genetic algorithm (evolutionary)
- Template-based prompt generation

#### 12. Integration Module (`mod.zig`)
**Purpose:** Export all evaluation frameworks
**LOC:** ~100

### Project Structure
```
agenkit-zig/
├── src/
│   └── evaluation/
│       ├── core.zig               (Phase 1.1)
│       ├── metrics.zig            (Phase 1.2)
│       ├── quality_metrics.zig    (Phase 1.3)
│       ├── recorder.zig           (Phase 1.4)
│       ├── regression.zig         (Phase 1.5)
│       ├── context_metrics.zig    (Phase 2.1)
│       ├── benchmarks.zig         (Phase 2.2)
│       ├── optimizer.zig          (Phase 2.3)
│       ├── ab_testing.zig         (Phase 2.4)
│       ├── bayesian_optimizer.zig (Phase 3.1)
│       ├── prompt_optimizer.zig   (Phase 3.2)
│       └── mod.zig                (Exports)
├── tests/
│   └── evaluation/
│       ├── test_core.zig
│       ├── test_metrics.zig
│       ├── test_quality_metrics.zig
│       ├── test_recorder.zig
│       ├── test_regression.zig
│       ├── test_context_metrics.zig
│       ├── test_benchmarks.zig
│       ├── test_optimizer.zig
│       ├── test_ab_testing.zig
│       ├── test_bayesian_optimizer.zig
│       └── test_prompt_optimizer.zig
└── examples/
    └── evaluation/
        ├── basic_metrics.zig
        ├── context_metrics.zig
        ├── benchmarks.zig
        ├── optimizer.zig
        ├── ab_testing.zig
        ├── bayesian_optimizer.zig
        └── prompt_optimizer.zig
```

#### Memory Management
```zig
// Allocator-based approach
pub fn init(allocator: std.mem.Allocator) !Self {
    return Self{
        .allocator = allocator,
        .history = std.ArrayList(OptimizationStep).init(allocator),
        // ...
    };
}

pub fn deinit(self: *Self) void {
    self.history.deinit();
    // Clean up other allocated resources
}
```

#### Error Handling
```zig
pub const EvaluationError = error{
    InvalidConfig,
    EmptyVariations,
    OptimizationFailed,
    StatisticalError,
};

pub fn optimize(self: *Self, n_iterations: usize) !OptimizationResult {
    if (n_iterations == 0) return EvaluationError.InvalidConfig;
    // ...
}
```

#### Comptime Features
Use comptime for type-safe configurations:
```zig
pub fn Optimizer(comptime T: type) type {
    return struct {
        const Self = @This();
        objective: *const fn (*const T) f64,
        search_space: SearchSpace(T),
        // ...
    };
}
```

#### Async/Await
Zig's async model for non-blocking operations:
```zig
pub fn optimize(self: *Self, n_iterations: usize) !OptimizationResult {
    var frame = async self.optimizeAsync(n_iterations);
    return await frame;
}
```

### Phase 1: Foundation (Weeks 1-3)
- [ ] Create `/src/evaluation` directory structure
- [ ] Implement `core.zig` with foundation types
- [ ] Implement `metrics.zig` for metrics collection
- [ ] Implement `quality_metrics.zig` for standard metrics
- [ ] Implement `recorder.zig` for session recording
- [ ] Implement `regression.zig` for regression detection
- [ ] Write tests for Phase 1 (~50+ tests)
- [ ] Create basic examples

### Phase 2: Advanced Features (Weeks 4-6)
- [ ] Implement `context_metrics.zig` for scale evaluation
- [ ] Implement `benchmarks.zig` with test suites
- [ ] Implement `optimizer.zig` with SearchSpace
- [ ] Implement `ab_testing.zig` with statistical tests
- [ ] Write tests for Phase 2 (~40+ tests)
- [ ] Create advanced examples

### Phase 3: Optimization (Weeks 7-9)
- [ ] Implement `bayesian_optimizer.zig` with acquisition functions
- [ ] Implement `prompt_optimizer.zig` with all strategies
- [ ] Write tests for Phase 3 (~30+ tests)
- [ ] Create optimization examples

### Integration & Documentation (Week 10)
- [ ] Implement `mod.zig` to export all modules
- [ ] Update `build.zig` for evaluation tests/examples
- [ ] Write comprehensive documentation
- [ ] Create evaluation README
- [ ] Verify cross-language compatibility

## Estimated Effort

**Total LOC:** ~12,000 LOC
- **Implementations:** ~6,300 LOC (12 modules)
- **Tests:** ~3,500 LOC (120+ tests)
- **Examples:** ~1,800 LOC (12 examples)
- **Documentation:** ~400 LOC (README, comments)

**Timeline:** 10-12 weeks for complete implementation

**Phases:**
1. Foundation: 3 weeks
2. Advanced: 3 weeks
3. Optimization: 3 weeks
4. Integration: 1 week

## Reference Implementations

Use these as reference (in order of recency):

1. **C++ (Most Recent, Best Reference):**
   - `/agenkit-cpp/src/evaluation/` - Implementation
   - `/agenkit-cpp/include/agenkit/evaluation/` - Headers
   - `/agenkit-cpp/tests/unit/` - Tests
   - **Status:** Complete (12/12), 12,583 LOC
   - **Approach:** Simplified algorithms, no heavy dependencies

2. **TypeScript:**
   - `/agenkit-ts/src/evaluation/` - Implementation
   - **Status:** Complete (12/12)
   - **Patterns:** Async/await patterns useful for Zig

3. **Go:**
   - `/agenkit-go/evaluation/` - Implementation
   - **Status:** Complete (12/12)
   - **Approach:** Simplified surrogate models, practical algorithms

4. **Python (Original):**
   - `/agenkit/evaluation/` - Reference
   - **Status:** Complete (12/12)
   - **Note:** Some features use sklearn (full GP) - use simplified approach

### Test Coverage Goals
- **Unit tests:** 120+ tests total
  - Phase 1: 50+ tests
  - Phase 2: 40+ tests
  - Phase 3: 30+ tests
- **Integration tests:** Cross-module workflows
- **Example tests:** Verify all examples compile and run

### Test Command
```bash
zig build test -Doptimize=ReleaseFast
```

### Example Command
```bash
zig build run-example -Dexample=bayesian_optimizer
```

### Must Have
- [ ] All 12 frameworks implemented
- [ ] 120+ tests passing (100% pass rate)
- [ ] 12+ working examples
- [ ] Comprehensive documentation
- [ ] Clean `zig build test` run
- [ ] No memory leaks (valgrind clean)
- [ ] Zig 100% evaluation parity achieved

### Nice to Have
- [ ] Performance benchmarks vs other languages
- [ ] Comptime optimization examples
- [ ] Cross-compilation verification
- [ ] WebAssembly target support

## Benefits

Completing Zig evaluation frameworks will:
1. ✅ Achieve 6-language evaluation parity (Python/Go/TS/C++/Rust/Zig)
2. 🚀 Enable systematic agent optimization in Zig
3. ⚡ Leverage Zig's performance advantages
4. 🔒 Provide memory-safe evaluation infrastructure
5. 🌐 Support Zig's excellent cross-compilation
6. 📊 Complete the evaluation toolkit across all languages

## Related Issues

- #146 - [C++] Implement All Evaluation Frameworks ✅ Complete
- #309 - [Rust] Complete Evaluation Frameworks (A/B Testing, Bayesian Optimization)
- Future: Cross-language evaluation server

### Why This is Worth Doing
- Zig is gaining traction for systems programming
- Excellent C interop enables unique use cases
- Performance characteristics rival C/C++
- Growing ecosystem for AI/ML applications

### Implementation Philosophy
- Follow Zig idioms (allocator patterns, error unions, comptime)
- Use simplified algorithms (match Go/C++ approach)
- Avoid heavy dependencies
- Prioritize clarity and correctness over optimization initially
- Profile and optimize hot paths after correctness established

## Priority

**Medium Priority** - Not blocking v0.46.0 release, but important for:
- Complete 6-language parity
- Zig ecosystem growth
- Cross-platform evaluation capabilities

Consider deferring to v0.47.0+ if other critical work takes precedence.
